Documentation/app-operations.md

```diff
@@ -0,0 +1,225 @@
+      Active: active (running) since Wed 2016-07-09 17:00:50 UTC; 8min 38s ago
+       Main PID: 6 (mysqld)
+          CGroup: /machine.slice/machine-rkt\x2d11efeefb\x2d2e8d\x2d47d0\x2da4ee\x2de1631327acac.scope/system.slice/mysql.service
+                     └─6 mysqld
+
+  Jul 15 17:01:03 rkt-$uuid mysql[6]: [Note]   - '::' resolves to '::';
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Note] Server socket created on IP: '::'.
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Note] InnoDB: Loading buffer pool(s) from /var/lib/mysql/ib_buffer_pool
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Warning] 'db' entry 'sys mysql.sys@localhost' ignored in --skip-name-resolve mode.
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Warning] 'proxies_priv' entry '@ root@localhost' ignored in --skip-name-resolve mode.
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Note] InnoDB: Buffer pool(s) load completed at 160713 17:01:03
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Warning] 'tables_priv' entry 'sys_config mysql.sys@localhost' ignored in --skip-name-resolve mode.
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Note] Event Scheduler: Loaded 0 events
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: [Note] mysqld: ready for connections.
+  Jul 13 17:01:03 rkt-$uuid mysql[6]: Version: '5.7.13'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server (GPL)
+```
+
+## rkt app exec
+Executes a command inside an application.
+
+```bash
+rkt app exec <pod-uuid> --app=<app-name> <arguments> -- <command> <command-arguments>
+```
+
+# Entrypoints
+
+In order to facilitate the app-level operations API, 4 new entrypoints are introduced.
+These entrypoints are stage1 binaries that are capable of communicating with the stage1 supervisor.
+
+For example, the `app start` entrypoint in default stage1 flavor sends a `start` signal
+(similar to `systemctl start`) for the service files of the application to the systemd which is the stage1 supervisor.
+
+## rkt app inject
+
+`coreos.com/rkt/stage1/app/inject`
+
+1. receive the prepared-app intended for injection consisting of it's rootfs, unit files, volumes, etc.
+2. `coreos.com/rkt/stage1/inject` entrypoint is resolved via annotations found within `/var/lib/rkt/pods/run/$uuid/stage1/manifest`.
+3. the entrypoint will be executed and passed a reference to the rootfs of the application that was prepared.
+
+The responsibility of this entrypoint is to receive a prepared app, copy its contents into stage2, 
+so that it is ready to be started by the stage1 supervisor (e.g. systemd for the default flavor).
+
+### Stage1 Default Flavor Example Workflow
+
+1. Configure the application service files given an application image/images.
+2. Copy the application rootfs to `/var/lib/rkt/pods/run/$uuid/stage1/rootfs/opt/stage2/`.
+3. Copy the service files of the application to the stage1 units directory `/usr/lib/systemd/system`.
+4. Start the injected application using the `app start` entrypoint.
+
+## rkt app eject
```

Comment:
I was filing an issue for that, but you beat me to it.

Such a pod wouldn't terminate unless killed.

For creation the boundary is not very strict because a pod without an app
is useless, so we can do a lazy pod creation where we would accumulate
configuration and then create the pod once the first application is
created, but we have to let an empty pod stay alive (perhaps specified via
`--allow-empty` flag when creating the pod) to allow the caller to stop and
start the same app, or we can add the additional `restart` api call to
prevent the RTT and the pod termination

On Fri, Jul 15, 2016 at 1:30 PM Alban Crequy <notifications@github.com>
wrote:

> In Documentation/app-operations.md
> <https://github.com/coreos/rkt/pull/2932#discussion_r70958427>:
>
> > +
> > +1. receive the prepared-app intended for injection consisting of it's rootfs, unit files, volumes, etc.
> > +2. `coreos.com/rkt/stage1/inject` <http://coreos.com/rkt/stage1/inject> entrypoint is resolved via annotations found within `/var/lib/rkt/pods/run/$uuid/stage1/manifest`.
> > +3. the entrypoint will be executed and passed a reference to the rootfs of the application that was prepared.
> > +
> > +The responsibility of this entrypoint is to receive a prepared app, copy its contents into stage2,
> > +so that it is ready to be started by the stage1 supervisor (e.g. systemd for the default flavor).
> > +
> > +### Stage1 Default Flavor Example Workflow
> > +
> > +1. Configure the application service files given an application image/images.
> > +2. Copy the application rootfs to `/var/lib/rkt/pods/run/$uuid/stage1/rootfs/opt/stage2/`.
> > +3. Copy the service files of the application to the stage1 units directory `/usr/lib/systemd/system`.
> > +4. Start the injected application using the `app start` entrypoint.
> > +
> > +## rkt app eject
>
> Is it allowed to remove all applications or should rkt enforce that there
> should be at least one app at all time?
>
> If a pod of zero apps is allowed, that might complicate the lifecycle
> management. When would such a pod terminates?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/rkt/pull/2932/files/713e7f8ac7a56987f294293cedd304c29f589ad7#r70958427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE1S5-6B1AwzJXVOqbU29osIl2CVDbnNks5qV29kgaJpZM4JMJi5>
> .
>